rust/src/producer.rs

```diff
@@ -0,0 +1,175 @@
+//! This module provides an abstraction over a Kafka producer in
+//! order to allow client code to instantiate the producer
+//! implementation they want without depending on the rdkafka
+//! ThreadedProducer.
+//!
+//! It also simplify unit tests.
+
+use rdkafka::config::ClientConfig as RdKafkaConfig;
+use rdkafka::producer::{BaseRecord, ThreadedProducer};
+use rdkafka::producer::{DeliveryResult, ProducerContext};
+use rdkafka::ClientContext;
+#[cfg(test)]
+use std::cell::RefCell;
+use std::collections::HashMap;
+#[cfg(test)]
+use std::rc::Rc;
+use thiserror::Error;
+
+/// This structure wraps the parameters to initialize a producer.
+/// This struct is there in order not to expose the rdkafka
+/// details outside.
+#[derive(Debug, Clone)]
+pub struct KafkaConfig {
+    config_map: HashMap<String, String>,
+}
+
+impl KafkaConfig {
+    pub fn new_producer_config(
+        bootstrap_servers: &str,
+        override_params: Option<HashMap<String, String>>,
+    ) -> Self {
+        let mut config_map: HashMap<String, String> = HashMap::new();
+        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.into());
+
+        let config = Self { config_map };
+
+        apply_override_params(config, override_params)
+    }
+}
+
+impl From<KafkaConfig> for RdKafkaConfig {
+    fn from(item: KafkaConfig) -> Self {
+        let mut config_obj = RdKafkaConfig::new();
+        for (key, val) in item.config_map.iter() {
+            config_obj.set(key, val);
+        }
+        config_obj
+    }
+}
+
+fn apply_override_params<V>(
+    mut config: KafkaConfig,
+    override_params: Option<HashMap<String, V>>,
+) -> KafkaConfig
+where
+    V: Into<String>,
+{
+    if let Some(params) = override_params {
+        for (param, value) in params {
+            config.config_map.insert(param, value.into());
+        }
+    }
+    config
+}
+
+struct CaptureErrorContext;
+
+impl ClientContext for CaptureErrorContext {}
+
+impl ProducerContext for CaptureErrorContext {
+    type DeliveryOpaque = ();
+
+    fn delivery(&self, _: &DeliveryResult, _delivery_opaque: Self::DeliveryOpaque) {
+        // TODO: Do something useful. AS of now guarantees of delivery
```

Comment:
Consider to add at least an error log through `log` or `tracing`. The `ThreadedProducer` only returns errors from its `send` method if the producer is in an unhealthy state at the time of emission. All other errors would be swallowed silently, which makes it harder to understand the impact during outages.

Reply:
I was planning to add some decent logging here. 
Do you have suggestion for a library that has to be used in both relay and symbolicator (which have different logging policies) ?
Should I create an abstract logger trait and let Relay/Symbolicator implement it ?

Reply:
The de-facto standard for new crates has become the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate. It allows you to track both spans and events (latter operating a lot like logs). The `tracing` crate itself is merely the facade and is inactive without subscribers, which would be registered by the application using your library. The application can then further customize _which_ spans and which events it subscribes to specifically from your library.

`tracing` also supports structured logging for errors. For example, [here](https://github.com/getsentry/relay/blob/75397010d18044a17759161e74475297fe9b0493/relay-kafka/src/producer/utils.rs#L20-L31) is the delivery callback we use in Relay. Note that `relay_log::error!` is a re-export of `tracing::error!` (:

```rust
    fn delivery(&self, result: &DeliveryResult, _delivery_opaque: Self::DeliveryOpaque) {
        if let Err((error, message)) = result {
            tracing::error!(
                // Sentry SDK convention: Using `error=` makes this a Sentry error event with stack trace
                error = error as &dyn Error,
                // Sentry SDK convention: This will become a field on the breadcrumb, or a context field on error events
                payload_len = message.payload_len(),
                // Sentry SDK convention: Using `tags.` makes this a tag on the logged error event
                tags.topic = message.topic(),
                // Sentry SDK convention: Breadcrumb or error event message
                "failed to produce message to Kafka (delivery callback)",
            );
        }
    }
```

In your tests, you will want to register a `tracing_subscriber` (separate crate for dev-dependencies) and call `.with_test_writer()`. This has to be called in _every_ test you want logs. Example from Relay is [here](https://github.com/getsentry/relay/blob/344e258b2900e871d56f4b3b7e9d339087fb5777/relay-log/src/test.rs#L15-L21), and I recommend this configuration:

```rust
    tracing_subscriber::fmt::fmt()
        .with_target(true)  // adds the origin module to log output
        .with_test_writer() // enables thread-local test mode
        .compact()          // prints a single line per log, which is better for CI
        .try_init()             // skips errors when 
        .ok();
```

In your examples, you can also register a subscriber. By default, it will print to stdout, so you may want to override this:

```rust
tracing_subscriber::fmt::layer()
        .with_writer(std::io::stderr)
        .with_target(true)
        .compact()  // or .pretty()
        .init();
```

This should pick up the `RUST_LOG` env variable by default to configure which log level is print at runtime.

--

Alternatives would be `log` or `slog`, which support just logging. `log` is still more widely adopted, but there are compatibility shims both ways. We exclusively use `tracing` in Symbolicator and Relay.

Reply:
Thanks.
Done